File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/Toast/Toast.kt
// The native Jetpack Compose + Material 3 Toast shared surface — a parity port of
// web/src/components/feedback/Toast.tsx. The web source is a `ToastProvider` that owns a transient
// queue of toasts and renders them bottom-anchored: each toast is one of four variants (success /
// error / info / warning) with a tone icon + tinted border, a title, an optional message, an optional
// action (a React Router `<Link to>` OR a `<button onClick>`), and a dismiss control; it announces
// assertively for `error` (`role="alert"`) and politely otherwise (`role="status"`), and its entry/exit
// animation collapses under reduced motion.
//
// This surface is the native equivalent. All state flows through the shared [ToastViewModel] over
// the [ToastController] seam (the surface-owned `ToastProvider` analogue) — the view performs NO HTTP
// and no timing. The faithful mapping of the web behaviour:
//   • `ToastProvider`'s `toasts` queue → the [ToastController], folded by the ViewModel into the stack;
//   • `useToast` / `useOptionalToast` → [requireToastController] / [LocalToastController];
//   • each variant's icon + `border-{tone}/30` → the shared `toneGlyph` + `toneColor` palette;
//   • the `t('a11y.dismissNotification')` label → the generated i18n catalog (P1/S10);
//   • the `<Link to>` / `<button onClick>` action → the host-routed [ToastAction.Navigate] / the
//     [ToastAction.Callback], invoked then dismissed by the ViewModel;
//   • the per-toast `setTimeout(dismiss, duration)` → the ViewModel auto-dismiss clock;
//   • the spring entry (y 20→0, scale .95→1) suppressed under reduced motion (web
//     `useMotionPreference().reduce`) → the native `rememberReducedMotion`-gated entry;
//   • `role="alert"`/`role="status"` → a merged assertive / polite Compose live region.
//
// States reproduced (the honest set for a notification primitive — see ToastModel): the populated stack
// (every tone + action + message branch, including the assertive `error` variant) and the empty queue
// (an invisible overlay, exactly what the web host renders with no toasts). There is no remote read, so
// no loading/stale/offline lifecycle is invented (covenant #9). The one-shot `view.opened` diagnostic
// (P1/S11) is emitted on first composition.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/shared-surfaces/Toast) cannot form a valid Kotlin package. `MatchingDeclarationName`
// is suppressed for the co-located stateless renderer + previews.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.toast

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import io.teslasync.android.R
import io.teslasync.android.components.feedback.toneColor
import io.teslasync.android.components.feedback.toneGlyph
import io.teslasync.android.components.motion.LocalReducedMotion
import io.teslasync.android.components.motion.rememberReducedMotion
import io.teslasync.android.components.ui.BodyText
import io.teslasync.android.components.ui.Button
import io.teslasync.android.components.ui.ButtonSize
import io.teslasync.android.components.ui.ButtonVariant
import io.teslasync.android.components.ui.Icon
import io.teslasync.android.components.ui.IconButton
import io.teslasync.android.components.ui.IconButtonVariant
import io.teslasync.android.components.ui.IconSize
import io.teslasync.android.components.ui.TeslaGlyphs
import io.teslasync.android.data.LocalDataContainer
import io.teslasync.android.ui.theme.TeslaSyncTheme
import io.teslasync.android.ui.theme.generated.Spacing
import io.teslasync.shared.core.diagnostics.Logger

/** Test tag for the host overlay container — present in every state (populated + empty). */
const val TOAST_HOST_TEST_TAG: String = "toast-host"

/** Test tag identifying one rendered toast card — used by the per-tone + a11y UI tests. */
const val TOAST_CARD_TEST_TAG: String = "toast-card"

private val TOAST_MAX_WIDTH: Dp = 380.dp
private const val BORDER_ALPHA = 0.30f
private const val CONTAINER_ELEVATION_DP = 3
private val ICON_TOP_NUDGE: Dp = 2.dp
private const val DESCRIPTION_MAX_LINES = 2
private const val VIEW_ARROW = " \u2192"

// ── Entry (web spring: y 20→0, scale .95→1; reduced motion → fade only) ─────────────────────────────
private const val ENTRY_DURATION_MS = 400
private const val ENTRY_SCALE_FROM = 0.95f
private val ENTRY_SLIDE_FROM: Dp = 20.dp

/**
 * `useOptionalToast` — the CompositionLocal a host provides the app-scoped [ToastController] through.
 * `null` when no host is mounted, exactly like the web `useOptionalToast` returning `null` outside a
 * `ToastProvider`. Use [requireToastController] for the throwing `useToast` variant.
 */
val LocalToastController: ProvidableCompositionLocal<ToastController?> = staticCompositionLocalOf { null }

/**
 * `useToast` — the [ToastController] from [LocalToastController], throwing if no [ToastHost] is mounted
 * (web `useToast` throws "must be used within ToastProvider"). For primitives that should degrade
 * gracefully when no host is present, read [LocalToastController] directly (the optional variant).
 */
@Composable
@ReadOnlyComposable
fun requireToastController(): ToastController =
    LocalToastController.current ?: error("requireToastController requires a ToastHost (LocalToastController not provided)")

/**
 * Stateful entry point bound to the shared [ToastController] — the faithful port of the web
 * `ToastProvider`'s rendered toast stack. Binds the [ToastViewModel], records the one-shot
 * `view.opened` diagnostic (P1/S11), collects the folded queue, and renders the stateless surface
 * bottom-anchored.
 *
 * @param controller the shared toast queue holder; a host builds one [DefaultToastController] at the
 *   app root and also provides it through [LocalToastController] so any screen can enqueue toasts.
 * @param modifier optional layout modifier for the overlay container.
 * @param onNavigate invoked with a route when a navigation action is tapped; the host routes it to the
 *   deep link (web `<Link to>`). Defaults to a no-op so the surface is safe to host before nav is wired.
 * @param logger the sanctioned redacting logger; defaults to the app's [LocalDataContainer].
 */
@Composable
fun ToastHost(
    controller: ToastController,
    modifier: Modifier = Modifier,
    onNavigate: (String) -> Unit = {},
    logger: Logger = LocalDataContainer.current.logger,
) {
    val viewModel: ToastViewModel =
        viewModel(
            key = ToastRegistration.ID,
            factory = ToastViewModel.factory(controller, logger),
        )
    LaunchedEffect(viewModel) { viewModel.recordViewOpened() }
    val state by viewModel.state.collectAsStateWithLifecycle()
    val reducedMotion = rememberReducedMotion()

    ToastHostContent(
        state = state,
        reducedMotion = reducedMotion,
        onAction = { message -> viewModel.invokeAction(message, onNavigate) },
        onDismiss = viewModel::dismiss,
        modifier = modifier,
    )
}

/**
 * Stateless renderer for the host overlay — the unit/UI-test and preview entry point. Anchors the toast
 * column bottom-end (web `fixed bottom-4 right-4`, RTL-aware), capped to [TOAST_MAX_WIDTH]. An empty
 * queue renders an invisible, zero-content overlay — the faithful web behaviour (the host shows nothing
 * with no toasts); the [TOAST_HOST_TEST_TAG] container is always present so the surface is locatable in
 * every state. [reducedMotion] gates the entry animation exactly as the web `useMotionPreference().reduce`
 * does.
 */
@Composable
fun ToastHostContent(
    state: ToastHostState,
    reducedMotion: Boolean,
    onAction: (ToastMessage) -> Unit,
    onDismiss: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier.fillMaxWidth().testTag(TOAST_HOST_TEST_TAG),
        contentAlignment = Alignment.BottomEnd,
    ) {
        Column(
            modifier = Modifier.widthIn(max = TOAST_MAX_WIDTH).padding(Spacing.md),
            verticalArrangement = Arrangement.spacedBy(Spacing.sm),
        ) {
            state.toasts.forEach { message ->
                ToastCard(
                    message = message,
                    reducedMotion = reducedMotion,
                    onAction = { onAction(message) },
                    onDismiss = { onDismiss(message.id) },
                )
            }
        }
    }
}

/**
 * One transient toast — the native mirror of a single web `<motion.div role=…>`: the tone icon on the
 * left, the title + optional message + optional action in the middle, and the dismiss control on the
 * right, bordered in the variant tint and wrapped in the matching (assertive / polite) live region. The
 * card fades/scales/slides in on first composition (a plain fade under reduced motion).
 */
@Composable
private fun ToastCard(
    message: ToastMessage,
    reducedMotion: Boolean,
    onAction: () -> Unit,
    onDismiss: () -> Unit,
) {
    val accent = toneColor(message.tone.toFeedbackTone())
    val entry = rememberToastEntry(reducedMotion)
    val liveRegionMode = if (message.isAssertive) LiveRegionMode.Assertive else LiveRegionMode.Polite

    Surface(
        modifier =
            Modifier
                .widthIn(max = TOAST_MAX_WIDTH)
                .testTag(TOAST_CARD_TEST_TAG)
                .graphicsLayer {
                    alpha = entry
                    scaleX = toastEntryScale(entry, reducedMotion)
                    scaleY = toastEntryScale(entry, reducedMotion)
                    translationY = (1f - entry) * if (reducedMotion) 0f else ENTRY_SLIDE_FROM.toPx()
                }.semantics(mergeDescendants = true) { liveRegion = liveRegionMode },
        shape = MaterialTheme.shapes.large,
        color = MaterialTheme.colorScheme.surface,
        contentColor = MaterialTheme.colorScheme.onSurface,
        tonalElevation = CONTAINER_ELEVATION_DP.dp,
        shadowElevation = CONTAINER_ELEVATION_DP.dp,
        border = BorderStroke(1.dp, accent.copy(alpha = BORDER_ALPHA)),
    ) {
        Row(
            modifier = Modifier.padding(Spacing.md),
            horizontalArrangement = Arrangement.spacedBy(Spacing.sm),
        ) {
            Icon(
                imageVector = toneGlyph(message.tone.toFeedbackTone()),
                contentDescription = null,
                size = IconSize.Sm,
                tint = accent,
                modifier = Modifier.padding(top = ICON_TOP_NUDGE),
            )
            ToastCardBody(message = message, onAction = onAction, modifier = Modifier.weight(1f))
            ToastDismissButton(onDismiss = onDismiss)
        }
    }
}

/** The title + optional message + optional action column (web toast body). */
@Composable
private fun ToastCardBody(
    message: ToastMessage,
    onAction: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
        Text(
            text = message.title,
            style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold),
            color = MaterialTheme.colorScheme.onSurface,
        )
        message.message?.let { body ->
            BodyText(
                text = body,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = DESCRIPTION_MAX_LINES,
            )
        }
        message.action?.let { action ->
            Button(
                label = toastActionLabel(action),
                onClick = onAction,
                variant = ButtonVariant.Ghost,
                size = ButtonSize.Sm,
            )
        }
    }
}

/** The dismiss control — the web `<button aria-label="Dismiss notification"><X/></button>`. */
@Composable
private fun ToastDismissButton(onDismiss: () -> Unit) {
    IconButton(
        imageVector = TeslaGlyphs.Close,
        contentDescription = stringResource(R.string.translation_a11y_dismissNotification),
        onClick = onDismiss,
        variant = IconButtonVariant.Standard,
        size = IconSize.Sm,
        tint = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}

/** The action's visible label — a navigation action carries the web "label →" arrow; a callback does not. */
private fun toastActionLabel(action: ToastAction): String =
    when (action) {
        is ToastAction.Navigate -> action.label + VIEW_ARROW
        is ToastAction.Callback -> action.label
    }

// ── Motion helpers ──────────────────────────────────────────────────────────────────────────────────

@Composable
private fun rememberToastEntry(reducedMotion: Boolean): Float {
    val progress = remember { Animatable(0f) }
    LaunchedEffect(reducedMotion) {
        progress.snapTo(0f)
        progress.animateTo(1f, tween(durationMillis = if (reducedMotion) 0 else ENTRY_DURATION_MS))
    }
    return progress.value
}

private fun toastEntryScale(
    entry: Float,
    reducedMotion: Boolean,
): Float = if (reducedMotion) 1f else ENTRY_SCALE_FROM + (1f - ENTRY_SCALE_FROM) * entry

// ── Previews (tooling-only; sample toasts are never shipped UI) ──────────────────────────────────────

private fun previewMessage(
    tone: ToastTone,
    title: String,
    message: String? = null,
    action: ToastAction? = null,
): ToastMessage =
    ToastMessage(
        id = "preview-${tone.name}",
        tone = tone,
        title = title,
        message = message,
        action = action,
    )

@Composable
private fun previewSurface(state: ToastHostState) {
    TeslaSyncTheme(dynamicColor = false) {
        CompositionLocalProvider(LocalReducedMotion provides true) {
            ToastHostContent(
                state = state,
                reducedMotion = true,
                onAction = {},
                onDismiss = {},
            )
        }
    }
}

@Preview(name = "Toast — success")
@Composable
private fun ToastSuccessPreview() {
    previewSurface(
        ToastHostState(
            listOf(previewMessage(ToastTone.Success, title = "Settings saved", message = "Your changes are live.")),
        ),
    )
}

@Preview(name = "Toast — error (assertive)")
@Composable
private fun ToastErrorPreview() {
    previewSurface(
        ToastHostState(
            listOf(previewMessage(ToastTone.Error, title = "Couldn't save", message = "Check your connection and retry.")),
        ),
    )
}

@Preview(name = "Toast — info with View link")
@Composable
private fun ToastInfoActionPreview() {
    previewSurface(
        ToastHostState(
            listOf(
                previewMessage(
                    tone = ToastTone.Info,
                    title = "Battery alert",
                    message = "State of charge dropped below 20%.",
                    action = ToastAction.Navigate(label = "View", route = "/battery"),
                ),
            ),
        ),
    )
}

@Preview(name = "Toast — warning with Undo")
@Composable
private fun ToastWarningActionPreview() {
    previewSurface(
        ToastHostState(
            listOf(
                previewMessage(
                    tone = ToastTone.Warning,
                    title = "Rule deleted",
                    action = ToastAction.Callback(label = "Undo", onInvoke = {}),
                ),
            ),
        ),
    )
}

@Preview(name = "Toast — stacked queue")
@Composable
private fun ToastStackPreview() {
    previewSurface(
        ToastHostState(
            listOf(
                previewMessage(ToastTone.Success, title = "Settings saved"),
                previewMessage(ToastTone.Info, title = "Export ready", message = "Your CSV is ready to download."),
                previewMessage(ToastTone.Error, title = "Sync failed"),
            ),
        ),
    )
}

@Preview(name = "Toast — empty (invisible host)")
@Composable
private fun ToastEmptyPreview() {
    previewSurface(ToastHostState.Empty)
}

File: apps/apple/TeslaSync/shared-surfaces/Toast.swift
//
//  Toast.swift
//  TeslaSync — P4 shared surface · 0144 · Toast (Apple)
//
//  The public API of the transient toast surface — the SwiftUI parity of the web `ToastProvider` in
//  `components/feedback/Toast.tsx`. The web source exposes three things: a context provider mounted once
//  (`ToastProvider`), a hook callers read to post toasts (`useToast`), and a non-throwing variant
//  (`useOptionalToast`). The native peers are: the ``ToastHost`` below (mounted once over the app content,
//  the `ToastProvider` parity — it injects the ``ToastCenter`` into the environment and overlays the toast
//  stack), `@Environment(ToastCenter.self)` (the `useToast()` parity — traps when no host is mounted, like
//  the hook throwing), and `@Environment(ToastCenter.self) var center: ToastCenter?` (the
//  `useOptionalToast()` parity — `nil` when no host is mounted). The host binds the ``ToastCenter`` for the
//  once-only `view.opened` telemetry (P1/S11) and overlays the floating stack bottom-trailing. No
//  networking, no Tailwind ports.
//
//  States (every branch the web source has renders — a transient client-only feedback layer has no fetch
//  lifecycle, so it carries no loading / stale / offline data states to fabricate):
//    • empty — no toasts → a transparent resting layer that passes touches straight through to the content
//      below (the web container's `pointer-events-none`), never a blank box.
//    • data  — one card per live toast, oldest-first, capped to five (web `toasts.map`).
//    • error — the `error`-kind card announces assertively (web `role="alert"`); the rest announce
//      politely (web `role="status"`).
//

import SwiftUI

// MARK: - ToastOverlay (web the fixed bottom-right toast container)

/// The floating toast stack — the native parity of the web fixed `bottom-4 right-4` container. Renders one
/// ``ToastRowView`` per live toast bottom-trailing, springs each in / out (fade-only under Reduce Motion),
/// and stays transparent to touches where there is no card so the content below remains interactive (web
/// `pointer-events-none` on the container, `pointer-events-auto` on each toast).
public struct ToastOverlay: View {
    private let center: ToastCenter
    private let onNavigate: (String) -> Void
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    public init(center: ToastCenter, onNavigate: @escaping (String) -> Void = { _ in }) {
        self.center = center
        self.onNavigate = onNavigate
    }

    public var body: some View {
        ZStack(alignment: .bottomTrailing) {
            Color.clear.allowsHitTesting(false)
            if !center.items.isEmpty {
                stack
            }
        }
        .animation(reduceMotion ? nil : .spring(response: 0.4, dampingFraction: 0.85), value: center.items.count)
        .accessibilityElement(children: .contain)
    }

    /// The data render — one card per live toast, oldest-first (web `toasts.map`).
    private var stack: some View {
        VStack(alignment: .trailing, spacing: TSSpacing.md) {
            ForEach(center.items) { item in
                ToastRowView(
                    item: item,
                    onNavigate: onNavigate,
                    onDismiss: { center.dismiss(id: item.id) }
                )
                .transition(transition)
            }
        }
        .padding(TSSpacing.lg)
        .frame(maxWidth: 420, alignment: .trailing)
    }

    private var transition: AnyTransition {
        if reduceMotion {
            return .opacity
        }
        return .asymmetric(
            insertion: .move(edge: .trailing).combined(with: .opacity),
            removal: .move(edge: .trailing).combined(with: .opacity)
        )
    }
}

// MARK: - ToastHost (web `ToastProvider`)

/// The toast host — the SwiftUI parity of the web `ToastProvider` mounted once near the top of the tree.
/// Wrap the app content in it once: it injects the ``ToastCenter`` into the environment (so descendants
/// post toasts via `@Environment(ToastCenter.self)`, the `useToast()` parity), overlays the floating stack,
/// and emits `view.opened` once on first appear. The default host binds the app-global ``ToastCenter/shared``
/// (the single-provider parity); previews and tests inject their own center. `onNavigate` is the embedder's
/// router for a toast action's navigation path (the native peer of react-router resolving `<Link to=>`).
public struct ToastHost<Content: View>: View {
    /// Diagnostics surface slug (P1/S11 `view.opened`).
    public static var surfaceSlug: String {
        ToastSurface.slug
    }

    private let center: ToastCenter
    private let onNavigate: (String) -> Void
    private let content: Content

    public init(
        center: ToastCenter = .shared,
        onNavigate: @escaping (String) -> Void = { _ in },
        @ViewBuilder content: () -> Content
    ) {
        self.center = center
        self.onNavigate = onNavigate
        self.content = content()
    }

    public var body: some View {
        content
            .environment(center)
            .overlay {
                ToastOverlay(center: center, onNavigate: onNavigate)
            }
            .onAppear { center.start() }
            .onDisappear { center.stop() }
    }
}

// MARK: - Host modifier (web wrapping the app in `<ToastProvider>`)

public extension View {
    /// Mounts the toast host over this view — the native peer of wrapping the app in `<ToastProvider>`.
    /// Descendants post toasts through `@Environment(ToastCenter.self)` (the `useToast()` parity); the
    /// floating stack overlays bottom-trailing. `center` defaults to the app-global shared store, and
    /// `onNavigate` resolves a toast action's navigation path (web `<Link to=>`).
    func toastHost(
        center: ToastCenter = .shared,
        onNavigate: @escaping (String) -> Void = { _ in }
    ) -> some View {
        ToastHost(center: center, onNavigate: onNavigate) { self }
    }
}
